### PR TITLE
Clean up handling of default urls

### LINKED ISSUE #4761: Hardcoding port in ws url
In https://github.com/jupyterlab/jupyterlab/blob/4310ef3e2594fd451e50c921d50d346926dbb64b/packages/services/src/serverconnection.ts#L206-L220, we hardcode in a reference to port 8888. That doesn't seem good. Also, should we be checking that baseurl is an absolute url?

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #4761.  Cleans up handling of `baseUrl` and `wsUrl` in both `PageConfig` and `ServerConnection`.

- urls are now normalized using a new `URLExt.normalize` function.
- Do *not* fall back on `http://localhost`
- Clean up the fallback logic for `wsUrl`
- `makeSettings` does not use the `baseUrl` parameter in `PageConfig.getWsUrl` since technically that parameter doesn't make sense there, but I did not want to break the API in `coreutils`

### CODE REVIEW

**Comment by blink1073** [comment]
cc @vidartf

**Comment by vidartf** [comment]
@blink1073 For me, the most interesting scenario is the behavior when there is not a `jupyter-config-data` element on the page. As far as I can tell, the changes for that scenario now are:
- `ServerConnection.defaultSettings` will have a blank string for `baseUrl`/`wsUrl` instead of having it be based on `location`.
- Calling `ServerConnection.makeSettings()` without supplying at least a `baseUrl` option will throw an error instead of using `location`.

This means that in order to make code that is agnostic about the page containing a `jupyter-config-data` element, my current code will have to change as follows:

```diff
export
function requestJsonPromise(url: string, argument: any): Promise<JSONObject> {
  let request = {
      method: 'POST',
      body: JSON.stringify(argument),
    };
+ let baseUrl = ServerConnection.defaultSettings.baseUrl || location.origin + '/';
+ let settings = ServerConnection.makeSettings({ baseUrl });
-  let settings = ServerConnection.makeSettings();
  return ServerConnection.makeRequest(url, request, settings)
    .then(handleError)
    .then((response) => {
      return response.json();
    });
}
```

Correct? Anything I missed?

**Comment by blink1073** [comment]
You'd need to add 8888 to the url get the old behavior.

**Comment by blink1073** [comment]
Oh wait, I see what you mean about using window.location as a default.

**Comment by blink1073** [comment]
@vidartf, `PageConfig.baseUrl` will now fall back on `location.origin`, meaning the only time you can't get away with not specifying a `baseUrl` is in `node`.

File: packages/coreutils/src/pageconfig.ts [hunk 1]
```diff
@@ -110,17 +110,18 @@ export namespace PageConfig {
...
   export function getBaseUrl(): string {
-    let baseUrl = getOption('baseUrl');
-    if (!baseUrl || baseUrl === '/') {
-      baseUrl =
-        typeof location === 'undefined'
-          ? 'http://localhost:8888/'
-          : location.origin + '/';
+    const baseUrl = URLExt.normalize(getOption('baseUrl'));
+    if (!baseUrl && typeof location !== 'undefined') {
+      return URLExt.normalize(location.origin);
```
Comment (vidartf) [inline packages/coreutils/src/pageconfig.ts]: What happens if you simply use `URLExt.normalize('/')` here, and skip the location check? Will this work on node as well?

Comment (blink1073) [reply packages/coreutils/src/pageconfig.ts]: It gives back `/` in node, which would go to the root of your server.  However, that doesn't help us get a `wsUrl`...

**Comment by blink1073** [comment]
@vidartf, thinking more about your use case, I made `wsUrl` optional altogether and allow `/` on `node` as the default `baseUrl`.

**Comment by blink1073** [comment]
Having a default `wsUrl` in node is odd anyway, since WebSockets are not built in 😜.

**Pull request merged**